### PR TITLE
Integrates proof generation and verification into wakunode2

### PR DESCRIPTION
**Pull request by staheri14**
An increment towards https://github.com/status-im/nim-waku/issues/714 (checkmarks the first three items on the task list of https://github.com/status-im/nim-waku/issues/714 )

Summary of the PR:
- Implements the full zkp proof generation and verification
- Integrates proof-verification into rln-relay topic validator
- Introduces non-spam proof struct `NonSpamProof` and its protobuf equivalence (the `NonSpamProof` replaces the byte sequence which was previously used to represent the proof)
- Adds multiple utility procs which essentially aim to encapsulate the type conversations and other complications of interacting with the Nim wrappers of rln lib
- Edits the interface of wakunode `publish` method. In the previous version, the `publish` proc would get a flag indicating whether rln-relay is enabled or not, according to this a non-spam proof would be created and attached to the supplied waku message. However, this means the content of the waku message gets altered inside the proc (which is unconventional and has consequences that would be difficult to deal with). I have deleted the proof generation part from `publish` and delegated it to the caller procs.  

# Next
- Integrating proof generation into wakunode2
- Extending the message verification proc to validate the epochs of incoming messages against the current epoch
- Persisting proofs of relayed messages in order to spot double signaling and enable slashing

### CODE REVIEW

File: tests/v2/test_wakunode.nim [hunk 1]
```diff
@@ -636,6 +687,108 @@ procSuite "WakuNode":
...
+      ## node1 publishes a message with an invalid non-spam proof, the message is then relayed to node2 which in turn 
+      ## attempts to verify the non-spam proof and fails hence does not relay the message to node3, thus the relayHandler of node3
+      ## never gets called
+      ## verification at node2 occurs inside a topic validator which is installed as part of the waku-rln-relay mount proc
+      await node1.publish(rlnRelayPubSubTopic, message)
+      await sleepAsync(2000.millis)
+
+      check:
+        # the relayHandler of node3 never gets called
+        (await completionFut.withTimeout(10.seconds)) == false
```
Comment (jm-clius) [inline tests/v2/test_wakunode.nim:787]: Is it necessary to wait a full `10` seconds? Just thinking that these long waits in the CI is starting to add up.

**Review by jm-clius — APPROVED** [review]
This lgtm! I'm assuming the API change is not a breaking change, as the parameter you removed has never been used externally?

File: tests/v2/test_wakunode.nim [hunk 2]
```diff
@@ -636,6 +687,108 @@ procSuite "WakuNode":
...
+                                contentTopic: contentTopic,  
+                                proof: nonSpamProof)
+
+
+      ## node1 publishes a message with an invalid non-spam proof, the message is then relayed to node2 which in turn 
+      ## attempts to verify the non-spam proof and fails hence does not relay the message to node3, thus the relayHandler of node3
+      ## never gets called
+      ## verification at node2 occurs inside a topic validator which is installed as part of the waku-rln-relay mount proc
+      await node1.publish(rlnRelayPubSubTopic, message)
+      await sleepAsync(2000.millis)
```
[the remark this answers is not in the document]
Comment (staheri14) [reply tests/v2/test_wakunode.nim:783]: True!

**Comment by staheri14** [comment]
> This lgtm! I'm assuming the API change is not a breaking change, as the parameter you removed has never been used externally?

Exactly, it was defaulted to false and was not (needed to be) set/used at all.

**Comment by jm-clius** [comment]
> > > I see weird windows build errors (ubuntu and mac are fine), while I am investigating the problem, would certainly appreciate your review!
> > 
> > 
> > I see similar errors on different PR. Likely unrelated to what you've done then :)
> 
> Do you suggest merging irrespective of the windows CI error? @jm-clius

Well, the `Test / windows` and `jenkins/prs/linux` failures are expected and [investigated  here](https://github.com/status-im/nim-waku/issues/738)
I'd say as long as the CI passes on every platform (macos, linux/ubuntu, windows) on _either_ the jenkins or GH workflow train it should be fine to merge. WDYT? As long as we capture each CI error in an issue (#738 could work) and determine that it's unrelated to the changes being merged.

File: tests/v2/test_waku_rln_relay.nim [hunk 3]
```diff
@@ -590,6 +592,22 @@ suite "Waku rln relay":
...
+    # create an RLN instance
+    var rlnInstance = createRLNInstance()
+    check:
+      rlnInstance.isOk == true
+    let rln = rlnInstance.value
+    
+    # prepare the input
+    # TODO should add support for arbitrary messages, the following input is artificial 
+    var hashInput : array[32, byte]
+    for x in hashInput.mitems: x= 1
```
Comment (oskarth) [inline tests/v2/test_waku_rln_relay.nim:605]:
```suggestion
    for x in hashInput.mitems: x = 1
```

File: tests/v2/test_waku_rln_relay.nim [hunk 4]
```diff
@@ -732,4 +750,141 @@ suite "Waku rln relay":
...
+                          shareY: shareY,
+                          nullifier: nullifier)
+      protobuf = nsp.encode()
+      decodednsp = NonSpamProof.init(protobuf.buffer)
+
+    check:
+      decodednsp.isErr == false
+      decodednsp.value == nsp
+
+  test "test proofVrfy and proofGen for a valid proof":
```
Comment (oskarth) [inline tests/v2/test_waku_rln_relay.nim:785]: proofVerify? Vrfy seems like a non-standard abbreviation :P

File: tests/v2/test_waku_rln_relay.nim [hunk 5]
```diff
@@ -732,4 +750,141 @@ suite "Waku rln relay":
...
+        member_is_added = rln.insertMember(memKeys.idCommitment)
+      else:
+        # create a new key pair
+        let memberKeys = rln.membershipKeyGen()
+        member_is_added = rln.insertMember(memberKeys.get().idCommitment)
+      # check the member is added
+      doAssert(member_is_added)
+
+    # prepare the message 
+    # TODO this message format is artificial (to bypass the Poseidon hasher issue)
```
Comment (oskarth) [inline tests/v2/test_waku_rln_relay.nim:811]: What Poseidon hasher issue? Maybe link to GH issue or describe it

File: tests/v2/test_wakunode.nim (hunk 1 above)
Comment (oskarth) [reply tests/v2/test_wakunode.nim:786]: > BTW, RLN-Relay tests are not yet effective in our CI checks :) they are conditioned to the RLN flag which is set to false by default

Can we instrument this somehow? Perhaps set RLN to true in test context or so?

File: waku/v2/node/wakunode2.nim [hunk 6]
```diff
@@ -289,14 +289,13 @@ proc unsubscribe*(node: WakuNode, request: FilterRequest) {.async, gcsafe.} =
   waku_node_filters.set(node.filters.len.int64)
 
 
-proc publish*(node: WakuNode, topic: Topic, message: WakuMessage,  rlnRelayEnabled: bool = false) {.async, gcsafe.} =
+proc publish*(node: WakuNode, topic: Topic, message: WakuMessage) {.async, gcsafe.} =
```
Comment (oskarth) [inline waku/v2/node/wakunode2.nim:292]: How come we are change the public API here? Is

> When rlnRelayEnabled is true, a zkp will be generated and attached to the message

No longer true? If so, how do we do or plan to do this now?

File: waku/v2/protocol/waku_message.nim [hunk 7]
```diff
@@ -23,7 +24,7 @@ type
     # the proof field indicates that the message is not a spam
     # this field will be used in the rln-relay protocol
     # XXX Experimental, this is part of https://rfc.vac.dev/spec/17/ spec and not yet part of WakuMessage spec
-    proof*: seq[byte]
+    proof*: NonSpamProof
```
Comment (oskarth) [inline waku/v2/protocol/waku_message.nim:27]: NonSpamProof...maybe we can have a different name?

rlnProof
NotSpamProof
rateLimitProof

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim [hunk 8]
```diff
@@ -3,19 +3,26 @@
...
 type RLN*[E] = pointer
 
 
-type IDKey* = array[32, byte]
-type IDCommitment* = array[32, byte]
-# represents a Merkle tree node which is the output of 
-# Poseidon hash function implemented by rln lib
-type MerkleNode* = array[32,byte] 
+type 
+  IDKey* = array[32, byte]
```
Comment (oskarth) [inline waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim:17]: Can we add a brief comment above what these are or represent? Alternatively, add link to spec/docs where this is elaborated on.

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim [hunk 9]
```diff
@@ -25,9 +32,22 @@ type MembershipKeyPair* = object
   # Poseidon hash function implemented in rln lib
   idCommitment*: IDCommitment 
 
-type WakuRLNRelay* = object 
+type NonSpamProof* = object
```
Comment (oskarth) [inline waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim:35]: Ditto above re docs

**Comment by jm-clius** [comment]
Btw, `jenkins/prs/linux` should now be fixed if you merge changes from `master` into this branch.
`Test / windows` is still failing: https://github.com/status-im/nim-waku/issues/738
Not sure why your `jenkins/prs/windows` job is failing, though - this is not a known issue.

File: tests/v2/test_waku_rln_relay.nim (hunk 4 above)
Comment (staheri14) [reply tests/v2/test_waku_rln_relay.nim:785]: It is very common in crypto papers :D (but I am going to change it)

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim (hunk 8 above)
Comment (staheri14) [reply waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim:17]: I have already explained these types in the `MembershipKeyPair` object definition (for each object field), will also include those comments here

**Comment by staheri14** [comment]
> Btw, `jenkins/prs/linux` should now be fixed if you merge changes from `master` into this branch. `Test / windows` is still failing: #738 Not sure why your `jenkins/prs/windows` job is failing, though - this is not a known issue.

I have merged the latest master, yet `jenkins/prs/linux` is not gone @jm-clius

File: tests/v2/test_wakunode.nim (hunk 1 above)
Comment (staheri14) [reply tests/v2/test_wakunode.nim:787]: I am not sure how to selectively enable rln tests i.e., active on GH CIs but not for local builds. I am going to track it down in this issue https://github.com/status-im/nim-waku/issues/741

File: tests/v2/test_waku_rln_relay.nim (hunk 4 above)
Comment (oskarth) [reply tests/v2/test_waku_rln_relay.nim]: oh til

File: tests/all_tests_v2.nim [hunk 10]
```diff
@@ -1,23 +1,23 @@
...
-  ./v2/test_jsonrpc_waku,
-  ./v2/test_peer_manager,
-  ./v2/test_web3, # TODO  remove it when rln-relay tests get finalized
-  ./v2/test_waku_bridge,
-  ./v2/test_peer_storage,
-  ./v2/test_waku_keepalive,
-  ./v2/test_migration_utils,
-  ./v2/test_namespacing_utils,
-  ./v2/test_waku_dnsdisc
+# import
```
Comment (oskarth) [inline tests/all_tests_v2.nim:1]: Tests needs to be enabled before merging

**Comment by jakubgs** [comment]
Please do not use the `Rerun` button, because it will cause the `Deps` stage to not run:
https://github.com/status-im/nim-waku/blob/165e23515807dee923569455fd40f5b7ac8fd225/ci/Jenkinsfile.prs#L44-L56
Which will cause this failure:
```
groovy.lang.MissingPropertyException: No such property: v1changed for class: groovy.lang.Binding
    at groovy.lang.Binding.getVariable(Binding.java:63)
    at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onGetProperty(SandboxInterceptor.java:271)
    at org.kohsuke.groovy.sandbox.impl.Checker$7.call(Checker.java:353)
    at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:357)
    at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:333)
    at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:333)
    at org.kohsuke.groovy.sandbox.impl.Checker.checkedGetProperty(Checker.java:333)
    at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.getProperty(SandboxInvoker.java:29)
    at com.cloudbees.groovy.cps.impl.PropertyAccessBlock.rawGet(PropertyAccessBlock.java:20)
    at WorkflowScript.run(WorkflowScript:74)
```

**Review by rshiv — APPROVED** [review]

**Review by rshiv — APPROVED** [review]

**Pull request merged**